clients/da-vinci-client/src/main/java/com/linkedin/davinci/consumer/VeniceAfterImageConsumerImpl.java

```diff
@@ -118,24 +118,40 @@ protected static void adjustSeekCheckPointsBasedOnHeartbeats(
       }
       PubSubPosition heartbeatTimestampPosition =
           consumerAdapter.getPositionByTimestamp(topicPartition, currentVersionTimestamp);
-      if (checkpoints.get(topicPartition.getPartitionNumber()) == null) {
-        LOGGER.warn("No EOP checkpoint found for partition: {}", topicPartition.getPartitionNumber());
+      if (checkpoints.get(topicPartition.getPartitionNumber()) == null && heartbeatTimestampPosition == null) {
+        LOGGER.warn(
+            "No EOP checkpoint OR heartbeat position found for partition: {} seeking to tail",
+            topicPartition.getPartitionNumber());
         checkpoints.put(
             topicPartition.getPartitionNumber(),
             new VeniceChangeCoordinate(
                 topicPartition.getPubSubTopic().getName(),
-                heartbeatTimestampPosition,
+                PubSubPosition.LATEST,
                 topicPartition.getPartitionNumber()));
-        continue;
-      }
-      PubSubPosition eopPosition = checkpoints.get(topicPartition.getPartitionNumber()).getPosition();
-      if (heartbeatTimestampPosition.comparePosition(eopPosition) > 0) {
+      } else if (checkpoints.get(topicPartition.getPartitionNumber()) == null) {
+        LOGGER.warn("No EOP checkpoint found for partition: {}", topicPartition.getPartitionNumber());
         checkpoints.put(
             topicPartition.getPartitionNumber(),
             new VeniceChangeCoordinate(
                 topicPartition.getPubSubTopic().getName(),
                 heartbeatTimestampPosition,
                 topicPartition.getPartitionNumber()));
+      } else if (heartbeatTimestampPosition == null) {
```

Comment:
Should we move this up? Otherwise we risk checkpointing   `null` heartbeatTimestampPosition in the preceding else if block?

Reply:
I think it should be ok as is.  The full if ladder is:

if(tcheckpoints.get(topicPartition.getPartitionNumber()) == null && heartbeatTimestampPosition == null) {}
else if(checkpoints.get(topicPartition.getPartitionNumber()) == null)
else if(heartbeatTimestampPosition == null)
else {}

There are only two booleans, so only 4 possible branches, so we cover them all.  If heartbeatPosition is null and EOP position is null, then we would have caught it in the first block (most specific condition should always go at the top of the if ladder, with broader statements following).

Reply:
Oh yeah you're right.